### DISCUSSION TITLE
Error Serializing ("[object Object]") cannot be serialized as JSON.

### DISCUSSION THREAD

**Discussion by Roland-Hufnagel** [neuefische/questions#296] [Web] [answered]
Beim Laden der Seite erhalte ich diesen Fehler: 
![Bildschirmfoto 2022-11-01 um 15 17 45](https://user-images.githubusercontent.com/111574974/199255018-533b4a2d-302c-44f9-b71e-70814b76ec17.png)

questions ist hierbei ein Array mit Objekten. Mein Schema: 
```
import mongoose from "mongoose";

const { Schema } = mongoose;

const surveySchema = new Schema({
  title: { type: String, required: true },
  description: { type: String },
  date: { type: Date },
  url: { type: String },
  questions: [
    {
      title: { type: String },
      type: { type: String },
      answers: [String],
    },
  ],
});
const SurveyModel =
  mongoose.models.SurveyModel ||
  mongoose.model("SurveyModel", surveySchema, "surveys");

export default SurveyModel;
```
und die getServerSideProps:
```
export async function getServerSideProps(context) {
  const { id } = context.params;
  const survey = await getSurveyById(id);
  return {
    props: {
      ...survey,
    },
  };
}
```

Komischerweise verschwindet der Fehler, wenn ich diesen Hack verwende:
```
export async function getServerSideProps(context) {
  const { id } = context.params;
  const survey = await getSurveyById(id);
  return {
    props: {
      ...survey,
      questions: JSON.parse(JSON.stringify(survey.questions)),
    },
  };
}
```
So funktioniert zwar alles wieder, aber ganz koscher kommt mir diese Lösung nicht vor. Kann mir jemand erklären, was hier schief läuft?

**Comment by Mark-Krensel** [comment]
Hello, I have the very same error, only that the parsing & stringifying doesn't seem to work for me.

I aleady tried to add the follwoing to the props:
```jsx
      feasTime: JSON.parse(JSON.stringify(days.feasTime)),
      height: JSON.parse(JSON.stringify(days.height)),
      weight: JSON.parse(JSON.stringify(days.weight)),
```

### Schema
```jsx
import mongoose from "mongoose";

const { Schema } = mongoose;
const weightSchema = new Schema({ type: Number });
const heightSchema = new Schema({ type: Number });
const feastTimeSchema = new Schema({ type: String });

const daySchema = new Schema(
  {
    date: { type: String, required: true },
    weight: [weightSchema],
    height: [heightSchema],
    feastTime: [feastTimeSchema],
  },
  { versionKey: false }
);

const Day = mongoose.models.Day || mongoose.model("Day", daySchema);

export default Day;
```

### getServerSideProps
```jsx
export async function getServerSideProps() {
  const days = await getAllDays();
  return {
    props: {
      days: days,
    },
  };
}
```

**Reply by Mark-Krensel** [reply]
So this seems to get me somewhere: 
```jsx
    props: { days: JSON.parse(JSON.stringify(days)) },
```

**Comment by ThomasF85** [comment, marked as answer]
The props you want to send from your server to the client (over getServerSideProps) are going to be transformed into the JSON format on the server and then sent to the client. The client receives the data in JSON form and transforms it back into JavaScript objects. This process is called serialization / deserialization and happens behind the scenes. A lot of things can go wrong here and some Objects cannot be correctly transformed back into JavaScript Objects.
Next.js is very strict here and gives you error when it is not sure how to serialize an Object. JSON.stringify tries much harder to serialize an Object. You have to make sure to only return serializable objects from getServerSideProps. You can transform your objects to be serializable like this:
```
products.map((product) => ({
    name: product.name,
    description: product.description,
    price: product.price,
    category: product.category.name,
  }));
```
This way the resulting products will only have serializable properties.
It is also okay for you to use JSON.stringify and JSON.parse.
